app/models/deletion_request.rb

```diff
@@ -0,0 +1,57 @@
+class DeletionRequest < ApplicationRecord
+  validates_presence_of :uid, :provider, :pid
+
+  # there can only be one entry with given provider + uid
+  validates_uniqueness_of :uid, scope: :provider
+
+  before_validation :set_pid
+
+  def run
+    associated_user&.destroy!
+  end
+
+  def deleted?
+    associated_user.nil?
+  end
+
+  def self.from_signed_fb(req)
+    data = DeletionRequest.parse_fb_request(req)
+    return unless data
+    DeletionRequest.create(provider: "facebook", uid: data["user_id"])
+  end
+
+  def self.parse_fb_request(req)
+    encoded, payload = req.split(".", 2)
+    decoded = Base64.urlsafe_decode64(encoded)
+    data = JSON.parse(Base64.urlsafe_decode64(payload))
+
+    # we need to verify the digest is the same
+    exp = OpenSSL::HMAC.digest("SHA256", ENV["FACEBOOK_APP_SECRET"], payload)
+    raise FailedAuthentication, "FB deletion callback called with invalid data" if decoded != exp
+
+    data
+  end
+
+  private
+
+  def associated_user
+    # more providers will be added
```

Comment:
How many more? That will affect how this logic grows, & maybe how many more `User#*_user_id` attributes you add.

Reply:
OK, is this a TBD?

Reply:
I'm not sure, but I would be surprised if there's more after this. @frett do you know? Can probably remove this comment now